### PR TITLE
Java compilation refactoring and fixes

### PR DESCRIPTION
**Pull request by eed3si9n**
This is sort of a continuation of #128

There has been weird indirection around how Java compilation is wired up probably to keep bincompat during 0.13. @Duhemm's #128 started to clean that up, but I ran into an issue when I tried to adjust sbt to actually use it because `JavaTools` is what's needed not `JavaCompiler`.
Making `IncrementalCompilerJavaTools` [extend `JavaCompiler`](https://github.com/sbt/zinc/pull/128/commits/76a1a172641f443b8a4d6b9a70bb624c55955f62#diff-e8d61b91c2bfc9b46d463ed2e85d8aebR50) I don't think was the right choice. Instead now `JavaTool`, `JavaTools`, `JavaCompiler`, and `Javadoc` are all under `xsbti` as Java interfaces.

During testing I've also caught a Java error parsing bug, which was committed separately - 09933e8

/review @dwijnand @Duhemm

### CODE REVIEW

File: internal/compiler-interface/src/main/java/xsbti/compile/JavaTool.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package xsbti.compile;
+
+import java.io.File;
+import xsbti.Logger;
+import xsbti.Reporter;
+
+/**
+ * An interface for on of the tools in the java tool chain.
```
Comment (dwijnand) [inline internal/compiler-interface/src/main/java/xsbti/compile/JavaTool.java]: s/on/one/

File: internal/compiler-interface/src/main/java/xsbti/compile/JavaTool.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * We assume the following is true of tools:
+ * - The all take sources and options and log error messages
```
Comment (dwijnand) [inline internal/compiler-interface/src/main/java/xsbti/compile/JavaTool.java]: s/The/They/

File: internal/compiler-interface/src/main/java/xsbti/compile/JavaTool.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * - They return success or failure.
+ */
+public interface JavaTool {
+  /**
+   * This will run a java compiler / or other like tool (e.g. javadoc).
```
Comment (dwijnand) [inline internal/compiler-interface/src/main/java/xsbti/compile/JavaTool.java]: I think we should drop the slash

Comment (eed3si9n) [inline internal/compiler-interface/src/main/java/xsbti/compile/JavaTool.java]: Sure. I just copy pasted from the existing comment including the above typos.

File: internal/compiler-interface/src/main/java/xsbti/compile/JavaTool.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import xsbti.Logger;
+import xsbti.Reporter;
+
+/**
+ * An interface for on of the tools in the java tool chain.
+ *
+ * We assume the following is true of tools:
+ * - The all take sources and options and log error messages
+ * - They return success or failure.
+ */
```
Comment (dwijnand) [inline internal/compiler-interface/src/main/java/xsbti/compile/JavaTool.java:17]: It looks like they also take a reporter. But also, why duplicate this info here?

File: internal/zinc-compile-core/src/main/scala/sbt/internal/inc/javac/JavaCompiler.scala [hunk 5]
```diff
@@ -134,10 +57,21 @@ object JavaCompiler {
...
+      }
+      val augmentedClasspath = if (cpOptions.autoBoot) classpath ++ Seq(scalaInstance.libraryJar) else classpath
+      val javaCp = ClasspathOptionsUtil.javac(cpOptions.compiler)
+      (new CompilerArguments(scalaInstance, javaCp))(Array[File](), augmentedClasspath, Some(target), options)
+    }
 }
 
 /** Interface we use to document java code. This is a tag over the raw JavaTool interface. */
-trait Javadoc extends JavaTool {}
+// trait Javadoc extends JavaTool {}
```
Comment (dwijnand) [inline internal/zinc-compile-core/src/main/scala/sbt/internal/inc/javac/JavaCompiler.scala]: I think we should remove this now

Comment (eed3si9n) [inline internal/zinc-compile-core/src/main/scala/sbt/internal/inc/javac/JavaCompiler.scala]: Yea. I think I forgot to delete that line.

**Comment by dwijnand** [comment]
Minor details. LGTM, nice work.

File: internal/compiler-interface/src/main/java/xsbti/compile/JavaTool.java (hunk 4 above)
Comment (eed3si9n) [inline internal/compiler-interface/src/main/java/xsbti/compile/JavaTool.java:17]: I can wordsmith around what's intended to communicate here.

**Pull request merged**